### PR TITLE
Resolve issues #50 and #49

### LINKED ISSUE #49: Clicking "Save" or "Save As..." and canceling, results in false "saved" state
**Describe as detailed as possible when it happens**
While having unsaved changes, click the "Save" or "Save As..." option (or using shortcuts) and closing the dialogs (without actually saving the file) program marks changes as saved, so closing the program does not trigger "Are you sure?" dialog.

**Potential fix***
The fastest solution would be to just check if the file was saved successfully, and set the value of `_unsavedDocumentModified` in `ViewModelMain` corresponding to the result.

### PR DESCRIPTION
**Pull request by uuksu**
Mainly resolves issue #50 by implementating comparision of original document and modified document. Now there is clear indication of when document has actually changed.

Also fixes #49 even though it was already once fixed. Change made in #54 does not really matter here anymore.

### CODE REVIEW

**Comment by flabbet** [comment]
Hi, thanks for the PR! I have one problem though, this solution is not very good, cloning bitmap is super-not-efficient. For example, create a canvas of size 4096x4096 and try to draw something. Almost Impossible. Even that PixiEditor is for pixel-arts, so usually smaller canvases are used, the goal is still to provide the highest performance no matter what the user wants.

**Comment by uuksu** [comment]
Thanks for the review!

I see what you mean by using the editor gets impossible with large canvases. I'm aware that bitmap cloning is slow but it should not be THAT slow and the clone method is only called when opening a new document (and it's pretty fast considering what it is doing).

The actual reason why it was slowing down was pretty interesting. Now that Layer implement IEquatable<Layer> I found out that in LayerChange constructor there is call to ObservableCollection.IndexOf method for determining in which layer the change happens. Before today I did not know that IndexOf method actually calls the Equals method in object if possible! As you can quess, it can cause a pretty much lag if this Equal method that compares layer pixel by pixel is called every single time user draws pixel to the screen.

I fixed this by changing the IndexOf method to old school for-loop that tries to find reference without actually executing the comparision code.

**Comment by flabbet** [comment]
I see, now it runs much faster, but can you give me an example of how your solution is better over the old one? I drew a few pixels, saved the file, drew pixel, tried to close - popup prompted, as expected, but when I erased this pixel and tried to close again, popup showed anyway, same when you create a new document and try to close it without drawing anything

**Comment by uuksu** [comment]
It was actually missing the case that when saving it stores the state of the saved document. Now it should work when you save, draw, erase and try to close and it does not prompt you.

I'm totally sure if you're saying that it should not prompt when creating a new document and trying to close with no changes? I think it should prompt in this case a there are unsaved changes even when there are no changes at all.

**Comment by flabbet** [comment]
No, I meant that when you open the program, create a new document, do nothing with it, and close the program, it still prompts that document was changed

**Comment by uuksu** [comment]
Exactly what I meant too and my opinion is that it should work that way as it is **unsaved change**. But if you feel like it should work differently then it can be changed to work differently.

**Comment by uuksu** [comment]
As you wish, now it's now prompting for changes if canvas is empty.

**Review by flabbet — COMMENTED** [review]
It works, but not perfectly, sometimes it compares correctly, sometimes not, I think this solution is a bit overkill for the task, the longer I think about it I am closer to closing the issue, simply because it's hard to make a better solution than it is now without performance cost or complexity. If you have any idea of how we could improve it, write it here, if not I'll close the issue. 
Also, I wrote down some thoughts about your code, you don't need to implement them, simply because I won't merge the current solution, but it's just information for you.

File: PixiEditor/Models/DataHolders/Document.cs [hunk 1]
```diff
@@ -285,6 +285,46 @@ public void CenterContent()
...
+        {
+            if (other == null)
+            {
+                return false;
+            }
+
+            if (other.Width != Width && other.Height != Height)
+            {
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (uuksu) [reply PixiEditor/Models/DataHolders/Document.cs:310]: Well, everyone has their own taste I guess. If you feel like this is better way to do it, then do it that way. I don't like it so I won't do it.

File: PixiEditor/Models/Controllers/BitmapManager.cs [hunk 2]
```diff
@@ -257,6 +259,22 @@ public static bool IsOperationTool(Tool tool)
         {
             return tool is BitmapOperationTool;
         }
+
+        public bool IsActiveDocumentChanged()
```
[the remark this answers is not in the document]
Comment (uuksu) [reply PixiEditor/Models/Controllers/BitmapManager.cs:263]: Better is quite subjective here.

**Comment by uuksu** [comment]
You didnt really specify when it does work and when it does not so how would I even have change to check what is problem with it? I think it works and it is good solution to the problem as it is more reliable than the original solution. Yes, there is performance cost, but if you want to compare it pixel by pixel then there isn't much to do else than maybe multithread the comparision. I don't think the performance cost is even huge here as I really can't see it in my machine.

But if you feel like closing this issue, then go ahead. It's your software and I can't make everybody happy.

**Comment by flabbet** [comment]
For example, create an empty canvas, let's say 32x32, save it as .pixi, draw a square 2x2 in the middle, next check if prompt shows, yes, click cancel, next erase these pixels and check again, it shouldn't prompt but it does.

**Comment by uuksu** [comment]
I'm not sure if this the same thing you mentioned, but it seems that the Layers Offset calculation is failing.

I don't quite understand what the Offset means in this contect and why it changes? Is is a bug that it does not change back?

**Comment by flabbet** [comment]
Layers are dynamic which means, they resize to its content, so when you draw a 2x2 square, the layer (and bitmap) will be 2x2 size, no matter the document size, offset is a set of coordinates, which is used to position the layer in the viewport. Here is the visual representation of how it works (Offset is the coordinate of a top-left pixel). https://twitter.com/flabbet/status/1274087236481568771

**Comment by uuksu** [comment]
Ah ok, thanks for explaining. If the layer does not disappear when the last pixel is erased, then there is unsaved change even when the user can't see anything. This could be dealt by not comparing the Offset but I don't know if it has some other side effects I'm no aware of.

**Comment by flabbet** [comment]
If you can deal with it and apply the ifs suggestion I mentioned, I'll merge your solution

**Comment by uuksu** [comment]
Removed the Offset comparision from Layer.

Like I said in [here](https://github.com/PixiEditor/PixiEditor/pull/58#discussion_r498942428), I don't agree with those suggestions, but it's your code and your rules so they are now implemented.

File: PixiEditor/Models/DataHolders/Document.cs [hunk 3]
```diff
@@ -285,6 +285,46 @@ public void CenterContent()
...
+                Layer otherLayer = other.Layers.FirstOrDefault(l => l.Name == layer.Name);
+                if (otherLayer == null)
+                {
+                    return false;
+                }
+
+                if (!layer.Equals(otherLayer))
+                {
+                    return false;
+                }
```
Comment (flabbet) [inline PixiEditor/Models/DataHolders/Document.cs:323]:
```suggestion
                if (otherLayer == null || !layer.Equals(otherLayer))
                {
                    return false;
                }
```

**Review by flabbet — CHANGES_REQUESTED** [review]
It works quite well now, but still, I managed to find a bug, draw a 2x2 square, save, remove the top 2 pixels, and draw 2 pixels at the bottom, so you will have still a 2x2 square but moved one pixel down. And try to close, nothing will prompt, As you said, because it doesn't compare offsets, it doesn't properly detect changes in some cases. I think an easy fix just would be to again, add offset comparison, the last bug will come back, but you can just check if it's 1x1 and empty.

**Comment by flabbet** [comment]
Closing due to inactivity

**Pull request closed without merging**